lib/ingest_api/structured.ex

```diff
@@ -0,0 +1,100 @@
+defmodule Logger.Backend.Humio.IngestApi.Structured do
+  @moduledoc """
+    This Ingest API implementation is for Humio's `Structured` API.
+    [Humio Documentation]: https://docs.humio.com/api/ingest/#structured-data
+  """
+  @behaviour Logger.Backend.Humio.IngestApi
+  alias Logger.Backend.Humio.IngestApi
+
+  @path "/api/v1/ingest/humio-structured"
+  @content_type "application/json"
+  @omitted_metadata [:time, :mfa, :report_cb]
+
+  @impl true
+  def transmit(%{
+        log_events: log_events,
+        host: host,
+        token: token,
+        client: client,
+        format: format,
+        metadata_keys: metadata_keys
+      }) do
+    headers = IngestApi.generate_headers(token, @content_type)
+    events = to_humio_events(log_events, format, metadata_keys)
+    {:ok, body} = encode_events(events)
+
+    client.send(%{
+      base_url: host,
+      path: @path,
+      body: body,
+      headers: headers
+    })
+  end
+
+  defp encode_events(events) do
+    Jason.encode([
+      %{
+        "events" => events
+      }
+    ])
+  end
+
+  defp to_humio_events(log_events, format, metadata_keys) do
+    log_events |> Enum.map(&to_humio_event(&1, format, metadata_keys))
+  end
+
+  defp to_humio_event(
+         %{timestamp: timestamp, metadata: metadata} = log_event,
+         format,
+         metadata_keys
+       ) do
+    # omit metadata for raw string, we add metadata as attributes instead
+    raw_string = IngestApi.format_message(log_event, format, [])
+    attributes = metadata |> IngestApi.take_metadata(metadata_keys) |> metadata_to_map()
+
+    %{
+      "rawstring" => raw_string,
+      "timestamp" => format_timestamp(timestamp),
+      "attributes" => attributes
+    }
+  end
+
+  defp metadata_to_map(metadata) do
+    with m <- Enum.into(metadata, %{}),
+         m <- Map.drop(m, @omitted_metadata),
+         {app, m} <- Map.pop(m, :application),
+         {module, m} <- Map.pop(m, :module),
+         {function, m} <- Map.pop(m, :function),
+         {file, m} <- Map.pop(m, :file),
+         {line, m} <- Map.pop(m, :line),
+         {pid, m} <- Map.pop(m, :pid),
+         {gl, m} <- Map.pop(m, :gl),
+         {crash_reason, m} <- Map.pop(m, :crash_reason),
+         {initial_call, m} <- Map.pop(m, :initial_call),
+         {registered_name, m} <- Map.pop(m, :registered_name) do
```

Comment:
Consider using `Map.split/2` here:
```suggestion
         {fixups, others} <- Map.split(m, [:application, :module, :function, :file, :line, :pid, :gl, :crash_reason, :initial_call, :registered_name]) do
```

Reply:
The other thing here is that unless you have cases you should bail from and return early, `with` is unnecessary. Instead what if you did something like:

```elixir
  @fixup_keys [:application, :module, :function, :file, :line, :pid, :gl, :crash_reason, :initial_call, :registered_name]
  defp metadata_to_map(metadata) do
    {fixups, rest} = 
      metadata
      |> Enum.into(%{}) # This could also be Map.new/1
      |> Map.drop(@omitted_metadata)
      |> Map.split(@fixup_keys)

    rest
    |> put_val(:app, fixups[:application])
   # ...
```

Reply:
Thanks, that makes sense.  Will make those changes :)

Reply:
Since we don't know what information could be inside the `rest` fields that were not splitted into fixups, it might be a good idea to use `nil_or_inspect` on each value